src/metabase/lib/remove_replace.cljc

```diff
@@ -304,6 +309,58 @@
                  next-stage-number))
         query))))
 
+(defn- prefix
+  [part whole]
+  (when (every? true? (map = part whole))
+    part))
+
+(defn- on-stage-path
+  "Given a Malli `error` in `query`, return the a path to the erroneous part of the query we can remove
+  to fix the problem or nil if there is no such path."
+  [query error]
+  (let [in (:in error)]
+    ;; We can only fix problems in stages.
+    (when (= (first in) :stages)
+      (let [stage-number (second in)
+            path-in-stage (nnext in)]
+        ;; The path should point at least to a top level clause in the stage,
+        ;; e.g., a specific expression or an order-by clause, and
+        ;; it should have a stage-path as its prefix.
+        (when-let [p (and (next path-in-stage)
+                          (some #(prefix % path-in-stage) (stage-paths query stage-number)))]
+          ;; We keep the prefix of `in` that's pointing to a specific element on the stage path.
+          ;; 2 accounts for [:stages stage-number] and 1 for the key of the element on the path.
+          (subvec in 0 (+ (count p) 2 1)))))))
+
+(mu/defn ^:private replace-expression-removing-erroneous-parts :- :metabase.lib.schema/query
+  [unmodified-query :- :metabase.lib.schema/query
+   stage-number     :- :int
+   target           :- :metabase.lib.schema.expression/expression
+   replacement      :- :metabase.lib.schema.expression/expression]
+  (mu/disable-enforcement
+    (loop [query (tweak-expression unmodified-query stage-number target replacement)]
+      (let [explanation (mc/explain :metabase.lib.schema/query query)
+            error-paths (->> (:errors explanation)
+                             (keep #(on-stage-path query %))
+                             distinct)]
+        (if (seq error-paths)
+          (recur (reduce (fn [q path]
+                           (try
+                             (remove-clause q (second path) (get-in q path))
+                             (catch #?(:clj Exception :cljs js/Error) e
+                               (let [{:keys [error join]} (ex-data e)]
+                                 (if (= error :metabase.lib.util/cannot-remove-final-join-condition)
+                                   ;; remove the dangling join
+                                   (remove-join q (second path) join)
+                                   (throw e))))))
```

Comment:
This is the third place we play this trick. I'm not sure it's really worth preventing the cascading removal.